databend-jdbc/src/main/java/com/databend/jdbc/DatabendConnection.java

```diff
@@ -610,47 +615,68 @@ public void accept(DatabendSession session) {
         setSession(session);
     }
 
+    /**
+     * Retry executing a query in case of connection errors. fail over mechanism is used to retry the query when connect error occur
+     * It will find next target host based on configured Load balancing Policy.
+     * @param sql The SQL statement to execute.
+     * @param attach The stage attachment to use for the query.
+     * @param e The exception that occurred during the initial query execution.
+     * @param retry The number of times the query has been retried.
+     * @return A DatabendClient instance representing the successful query execution.
+     * @throws SQLException If the query fails after retrying the specified number of times.
+     * @see DatabendClientLoadBalancingPolicy
+     */
+    DatabendClient failover(String sql, StageAttachment attach, Exception e, int retry) throws SQLException {
+        if (e != null && !(e.getCause() instanceof ConnectException)) {
+            throw new SQLException("Error start query: " + "SQL: " + sql + " " + e.getMessage() + " cause: " + e.getCause(), e);
+        }
+        if (retry > this.getFailOver()) {
+            throw new SQLException("Error executing query: " + "SQL: " + sql + " " + e.getMessage() + " cause: " + e.getCause(), e);
+        }
+        try {
+//            if (!inActiveTransaction()) {
+//                this.routeHint = randRouteHint();
+//            }
+            // configure query and choose host based on load balancing policy.
+            ClientSettings.Builder sb = this.makeClientSettings();
+            if (attach != null) {
+               sb.setStageAttachment(attach);
+            }
+            ClientSettings s = sb.build();
+            logger.log(Level.FINE, "retry " + retry + " times to execute query: " + sql + " on " + s.getHost());
+            return new DatabendClientV1(httpClient, sql, s, this);
+        } catch (RuntimeException e1) {
+            return failover(sql, attach, e1, retry + 1);
```

Comment:
we can consider replacing this resursive call to a for loop

and catching RuntimeException might will not only catch the ConnectionError as we expected.